### PR TITLE
Add Federated Repositories

### PR DESCRIPTION
**Pull request by feliu**
Add federated repositories resource type.  This PR is a first draft, I still need to implement the attribute `members` in the resource `artifactory_federated_repository`.

Context: https://www.jfrog.com/confluence/display/JFROG/Repository+Configuration+JSON#RepositoryConfigurationJSON-FederatedRepository

Tested with Artifactory (SaaS) version 7.19.6.

### CODE REVIEW

**Comment by feliu** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by feliu** [comment]
recheckcla

File: pkg/artifactory/resource_artifactory_federated_repository.go [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+	repo.SuppressPomConsistencyChecks = d.getBoolRef("suppress_pom_consistency_checks", false)
+	repo.XrayIndex = d.getBoolRef("xray_index", false)
+	repo.ForceNugetAuthentication = d.getBoolRef("force_nuget_authentication", false)
+
+	return repo
+}
+
+func resourceFederatedRepositoryCreate(d *schema.ResourceData, m interface{}) error {
+	c := m.(*ArtClient).ArtOld
+	fmt.Println("Hello")
```
Comment (chb0github) [inline pkg/artifactory/resource_artifactory_federated_repository.go:183]: remove

File: pkg/artifactory/resource_artifactory_federated_repository.go [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+		return err
+	}
+
+	d.SetId(*repo.Key)
+	return resourceFederatedRepositoryRead(d, m)
+}
+
+func resourceFederatedRepositoryRead(d *schema.ResourceData, m interface{}) error {
+	c := m.(*ArtClient).ArtOld
+	fmt.Println("Hello")
```
Comment (chb0github) [inline pkg/artifactory/resource_artifactory_federated_repository.go:197]: remove

File: pkg/artifactory/resource_artifactory_federated_repository.go [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+		logError(d.Set("handle_releases", repo.HandleReleases))
+		logError(d.Set("handle_snapshots", repo.HandleSnapshots))
+		logError(d.Set("checksum_policy_type", repo.ChecksumPolicyType))
+		logError(d.Set("max_unique_snapshots", repo.MaxUniqueSnapshots))
+		logError(d.Set("snapshot_version_behavior", repo.SnapshotVersionBehavior))
+		logError(d.Set("suppress_pom_consistency_checks", repo.SuppressPomConsistencyChecks))
+		logError(d.Set("xray_index", repo.XrayIndex))
+		logError(d.Set("force_nuget_authentication", repo.ForceNugetAuthentication))
+
+		if hasErr {
```
Comment (chb0github) [inline pkg/artifactory/resource_artifactory_federated_repository.go:231]: this should be checked at line 205

File: pkg/artifactory/resource_artifactory_federated_repository.go [hunk 4]
```diff
@@ -0,0 +1,277 @@
+package artifactory
+
+import (
+	"context"
+	"fmt"
+	"net/http"
```
Comment (chb0github) [inline pkg/artifactory/resource_artifactory_federated_repository.go:6]: I'm not seeing where this gets used

File: pkg/artifactory/provider.go [hunk 5]
```diff
@@ -2,6 +2,9 @@ package artifactory
 
 import (
 	"fmt"
+	"net/http"
```
Comment (chb0github) [inline pkg/artifactory/provider.go:5]: You just reordered the imports?

File: pkg/artifactory/resource_artifactory_federated_repository.go (hunk 4 above)
Comment (feliu) [reply pkg/artifactory/resource_artifactory_federated_repository.go:6]: The `net/http` is used in line 200: `if resp.StatusCode == http.StatusNotFound {`

File: pkg/artifactory/provider.go (hunk 5 above)
Comment (feliu) [reply pkg/artifactory/provider.go:5]: vscode did it, let me amend it :)

Comment (feliu) [reply pkg/artifactory/provider.go:5]: I applied `gofmt` to have consistent formatting.

File: pkg/artifactory/resource_artifactory_federated_repository.go (hunk 3 above)
Comment (feliu) [reply pkg/artifactory/resource_artifactory_federated_repository.go:231]: I amend it already 👍

File: pkg/artifactory/resource_artifactory_federated_repository.go (hunk 1 above)
Comment (feliu) [reply pkg/artifactory/resource_artifactory_federated_repository.go]: done

**Comment by chb0github** [comment]
This is listed as a draft. Is it still active?

**Comment by feliu** [comment]
Hi @chb0github, I want to test it with my both Artifactory Cloud instances. I opened a ticket to JFrog Support, to help me on setting up a circle of trust between two Artifactory Cloud instances, and I'm still waiting. Once, I get a response from JFrog, I will be able to do a full test and complete this PR. The problem is that there is no documentation about setting a circle of trust between SaaS instances.

**Comment by feliu** [comment]
I just received a response from JFrog. in this case, I will be able to work with this task on Monday.

**Comment by feliu** [comment]
Depends on: https://github.com/atlassian/go-artifactory/pull/34

**Comment by cawilson** [comment]
> Depends on: [atlassian/go-artifactory#34](https://github.com/atlassian/go-artifactory/pull/34). Do you know if the repository atlassian/go-artifactory is actively maintained, @chb0github?

From what I can see, Atlassian hasn't touched the repository for 2 years, so I would say they have abandoned it. There are a lot of crucial PR's opened against it, which are required to move this Terraform provider forward and I think, barring any issues, Jfrog needs to take a fork and help drive it forward or replace it with there own API library.

We're really (REALLY) keen for the federated repository support (as an Enterprise+ user) and the lack of support is blocking us from converting local repositories. I feel this PR will be blocked until the API library issues can be sorted. You could write your own API library and pull that in (like what was done for the Xray resources) but I feel this is just going to cause a mass of unnecessary complication/fragmentation.

**Comment by chb0github** [comment]
@cawilson and @feliu 

We have [our own Go client](https://github.com/jfrog/jfrog-client-go) so we definitely won't be forking this one. We've inherited enough from atlassian. We are using it a little.

If you are in need of a test server, let me know here or on gitter.

If you want to get some traction on this I would recommend raising a legit [support ticket here](https://jfrog.com/support/). 

My time is entirely commandeered. I can do PRs, assist with resources, etc. Sorry I can't do more.

**Comment by cawilson** [comment]
> @cawilson and @feliu
> 
> We have [our own Go client](https://github.com/jfrog/jfrog-client-go) so we definitely won't be forking this one. We've inherited enough from atlassian. We are using it a little.
> 
> If you are in need of a test server, let me know here or on gitter.
> 
> If you want to get some traction on this I would recommend raising a legit [support ticket here](https://jfrog.com/support/).
> 
> My time is entirely commandeered. I can do PRs, assist with resources, etc. Sorry I can't do more.

@feliu and @chb0github

Yeah, I thought there was one already. I think one of the biggest things to do, in order to unlock and drive this provider forward, would be to completely remove and replace the Atlassian API library with the Jfrog one. Then we will be able to gain more traction in changes to the API and then changes to this Terraform provider.

I'll see if I can find any time to give it a go, but no guarantees...

**Comment by cawilson** [comment]
@feliu and @chb0github 

I have been looking into all this and started attempts at changing the library across to https://github.com/jfrog/jfrog-client-go but realised the work is a bit large and there are a few key things missing from Jfrog Library that need to be solved. So I'm attempting to only tackle `federated` repositories first by adding them to the Jfrog Go library (see https://github.com/jfrog/jfrog-client-go/pull/375) and then we should be able to use that instead for the federated repository resources. It means bringing in a third library but I think it can be reduced to the Jfrog one bit by bit

**Comment by chb0github** [comment]
@cawilson - I agree with you. Completely moving away from anything atlassian at this point would be ideal. Also, solidifying the HTTP client(s) to the jfrog-go one.

**Comment by chb0github** [comment]
I am in the process right now of pitching atlassian. But, I am seriously reconsidering using our own client. I am going to close this for now. Feel free to reopen when you have something ready

**Pull request closed without merging**